freezegun/api.py

```diff
@@ -230,35 +230,21 @@ def __call__(self, func):
         return self.decorate_callable(func)
 
     def decorate_class(self, klass):
-        klass_mro = klass.mro()
-        for attr in dir(klass):
-            if attr.startswith("_"):
-                continue
+        seen = set()
+        for base_klass in klass.mro():
+            for (attr, attr_value) in base_klass.__dict__.items():
+                if attr.startswith('_') or attr in seen:
+                    continue
+                seen.add(attr)
 
-            attr_value = getattr(klass, attr)
-            if not callable(attr_value):
-                continue
+                if not callable(attr_value) or inspect.isclass(attr_value):
+                    continue
 
-            for base_klass in klass_mro:
                 try:
-                    attr_descriptor = base_klass.__dict__[attr]
-                except KeyError:
-                    pass
-                else:
-                    break
-            else:
-                # don't sweat the attribute magic
-                attr_descriptor = None
-
-            # Skip patching classmethods and staticmethods
-            if isinstance(attr_descriptor, (classmethod, staticmethod)):
```

Comment:
Right. All I've done in this revision is to refactor it such that, rather than get the class's internal namespace more than once, (once with `dir()`, and then iterating over the MRO for each name), it relies on only one loop over the MRO, with a `set` to keep track of names we've already seen (because they were overwritten in the subclass). This is basically what `dir()` was doing for us in the first place, (excluding classes with exotic metaclasses, which would cause problems anyway).

I started looking through the MRO, and each of its classes' `__dict__`s, in the first place, to avoid having to guess which class attribute values are proper functions (which act as instance methods), and which are instances of `classmethod` and `staticmethod`. (For that matter, this should now exclude properties as well, which makes sense.) Because rather, these are implemented in the class `__dict__` as non-callable descriptors, (which wrap functions) &ndash; and so the `not callable(attr_value)` [line](https://github.com/spulec/freezegun/pull/88/files#diff-80ef255f3bb85cfd652668d719317c10R240) can now exclude these without further inspection. So, freezegun will only wrap undecorated functions, (and callable objects, unless we start checking with `inspect.isfunction`).